src/target/stm32l4.c

```diff
@@ -141,61 +243,62 @@ static void stm32l4_add_flash(target *t,
 	target_add_flash(t, f);
 }
 
-enum ID_STM32L4 {
-	ID_STM32L43  = 0x435, /* RM0394, Rev.4 */
-	ID_STM32L45  = 0x462, /* RM0394, Rev.4 */
-	ID_STM32L41  = 0x464, /* RM0394, Rev.4 */
-	ID_STM32L47  = 0x415, /* RM0351, Rev.5 */
-	ID_STM32G07  = 0x460, /* RM0444/454, Rev.1 */
-	ID_STM32L49  = 0x461, /* RM0351, Rev.5 */
-	ID_STM32L4R  = 0x470, /* RM0432, Rev.5 */
-};
-
 static bool stm32l4_attach(target *t)
 {
 	if (!cortexm_attach(t))
 		return false;
 
-	bool dual_bank = false;
-	uint32_t idcodereg = STM32L4_DBGMCU_IDCODE_PHYS;
-	uint32_t size = 0;
-
-	switch(t->idcode) {
-	case ID_STM32L47:
-	case ID_STM32L49:
-	case ID_STM32L4R:
-		dual_bank = true;
-		break;
-	case ID_STM32G07:
-		idcodereg = STM32G0_DBGMCU_IDCODE_PHYS;
-		break;
-	}
+	/* Retrive chip information, no need to check return */
+	struct stm32l4_info const *chip = stm32l4_get_chip_info(t->idcode);
+
+
+	uint32_t idcodereg = (chip->family == FAM_STM32G0x)
+				     ? STM32L4_DBGMCU_IDCODE_PHYS
+				     : STM32G0_DBGMCU_IDCODE_PHYS;
+
 
 	/* Save DBGMCU_CR to restore it when detaching*/
 	uint32_t dbgmcu_cr = target_mem_read32(t, DBGMCU_CR(idcodereg));
 	t->target_storage = dbgmcu_cr;
+
 	/* Enable debugging during all low power modes*/
 	target_mem_write32(t, DBGMCU_CR(idcodereg), DBGMCU_CR_DBG_SLEEP | DBGMCU_CR_DBG_STANDBY | DBGMCU_CR_DBG_STOP);
 
-	size = (target_mem_read32(t, FLASH_SIZE_REG) & 0xffff);
-	if (dual_bank) {
-		uint32_t options =  target_mem_read32(t, FLASH_OPTR);
-		if (t->idcode == ID_STM32L4R) {
-			/* rm0432 Rev. 2 does not mention 1 MB devices or explain DB1M.*/
-			if (options & OR_DBANK) {
-				stm32l4_add_flash(t, 0x08000000, 0x00100000, 0x1000, 0x08100000);
-				stm32l4_add_flash(t, 0x08100000, 0x00100000, 0x1000, 0x08100000);
-			} else
-				stm32l4_add_flash(t, 0x08000000, 0x00200000, 0x2000, -1);
+
+	/* Free previously loaded memory map */
+	target_mem_map_free(t);
+
+	/* Add RAM to memory map */
+	if (chip->family == FAM_STM32G0x) {
+		target_add_ram(t, 0x20000000, chip->sram1 << 10);
+	} else {
+		target_add_ram(t, 0x10000000, chip->sram2 << 10);
+		/* All L4 beside L47 alias SRAM2 after SRAM1.*/
+		uint32_t ramsize = (t->idcode == ID_STM32L47)?
+			chip->sram1 : (chip->sram1 + chip->sram2 + chip->sram3);
+		target_add_ram(t, 0x20000000, ramsize << 10);
+	}
+
+	/* Add the flash to memory map. */
+	/* TODO: This should really be more data driven... */
```

Comment:
What do you mean bt more data driven?

Reply:
I really don't like the jumble of nested ifs, I would prefer something that takes the right action with fewer conditions (at the expense of more data, yes).
Maybe my working with the maintenance of SW with a Mc Cabe complexity of 183 has tainted me...
I've seen other todos here and there, but if you think this is not worth the reminder, I'll remove it.